Server endpoint for campaign stats

andioneto opened the pull request:
### server endpoint: /campaigns/:id/metrics
### controller: getCampaignMetrics
### new services:
- **getCallsByCampaignId:** [campaigns service] will fetch campaign by id with related call models
- **fetchResponsesByQuestionId:** [responses service] will fetch responses with specific question id

### if no errors will return an object ("metrics") with the following structure:
  1. status_distribution - object with keys as statuses, values as number of attempts
  2. outcome_distribution - like ^ but with outcomes
  3. responses_stats - list of response stat objects
   -question_id
    -response_distribution - object with keys as answers and values as count of answers
## notes:
If no call statuses, call outcomes, or question responses exist, default is set to 0.
**paragraph style responses are currently omitted as data vis for that seems out of the scope of our mvp**
**controller helpers folder will be reworked when controllers are cleaned up, leaving it here for now since they're going to change very soon**

## testing:
new service tests currently exist
if you have changed the campaigns service tests, there will be merge conflicts when you pull this in
tests for helper functions to come in a small follow up ticket

### example:
<img width="409" alt="screen shot 2017-09-19 at 10 51 38 pm" src="https://user-images.githubusercontent.com/18120906/30628778-46e1de62-9d8d-11e7-9089-cac9477d943d.png">

**helper functions**:
  - splitResponses
    + split responses into an array of response strings
    + [params: responseString (string)]
  - checkCampaignNotDraft
    + goal: throw error if metrics request is made for a campaign that is still in 'draft'
    + [params: campaignId (int), res (response obj)]
  - createStatusMap
    + goal: create map of statuses to frequency for the campaign (initialized to 0); mapped from 'status_distribution' prop in the metrics object
    + [params: metricsObj (object)]
  - createOutcomeMap
    + goal: create map of outcomes to frequency for the campaign (initialized to 0); mapped from 'outcome_distribution' prop in the metrics object
    + [params: metricsObj (object)]
  - createResponseMap
    + goal: create map of responses to specific questions, to frequency for the campaign (initialized to 0); mapped from ' response_stats' prop in the metrics object
    + props are question ids, which map to another object that represents the frequency of each reponse
  - countStatusAndOutcomeFrequency
    + goal: iterate over calls and 'count' the frequencies of each outcoome/status
    + [params: callsArray (array of call models/ objs)]
  - countResponseFrequency
    + goal: count the frequency of responses for a specific questions within a specific campaign
    + [params: responsesArray(array), metricsObj(obj), questionId(int)]
    + first splits the serialized repsonses into sub-arrays, then performs a shallow flatten, then counts freq.
  - handleQuestionResponses
    + goal: promisifies fetching responses by question_Id and call_id in a campaign (left join that returns response, call_id, question_id

andioneto committed:
started on metrics controller
added services: fetchResponseByQuestionId (question service), getCallsByCampaignId (campaign service)
added route for metrics: campaigns/:id/metrics

todo: implement promise.all for parallel query
currently a 'semi-naive' solution as a baseline

andioneto committed:
added Promise.all to handle fetching question responses and update the metrics object

TODO: add in responses statuses (mostly for errors)
should only be active for campaigns that are not in draft

andioneto committed:
Merge branch 'master' of https://github.com/Twilio-org/phonebank into ao-metrics

andioneto committed:
added in error and success status codes/messages

next:
figure out some basic testing for helper functions... at least the mapping functions

andioneto committed:
fixed a small bug with the check campaginDraft function
forgot to pass in the response object as an argument to the helper function

andioneto committed:
tests for response service fetchResponsesByQuestionId are done
next testing for new campain service getCallsByCampaignId

andioneto committed:
forgot to include done()

andioneto committed:
done with response testing
next is campaigns testing

andioneto committed:
finished with testing
going to look one more time at the map I'm sending back, will separate out all the answers from multi select questions

andioneto committed:
last thing to do:
use a dictionary of all the potential outcomes and statuses, default them to 0;

andioneto committed:
reworked to include all options for responses, statuses, and outcomes.
if none of any type are present, they map to a zero value, otherwise they map to > 0;
I removed paragraph type responses, as this seems like something we can't really support with basic data vis.

andioneto committed:
Merge branch 'master' of https://github.com/Twilio-org/phonebank into ao-metrics

andioneto committed:
fixed a slight problem with the responses service

will leave the old service there for now until I change the tests, as this will break the tests

created a controller helper folder for controller helpers, moved my new helpers as well as my helpers for campaigns into it

andioneto committed:
removed unnecessary exports in metrics helper file

andioneto committed:
updated test and removed old serivce from responses

andioneto committed:
bc im a derp and forgot to change the test:sever command back

andioneto committed:
another derp, fixed a broken test

andioneto committed:
Merge branch 'master' of https://github.com/Twilio-org/phonebank into ao-metrics

allieglotfelty commented on server/controllers/controller_helpers/metrics.js:
```diff
@@ -0,0 +1,115 @@
+  return callObjArray.map(call => call.attributes.id);
+}
+
+export function checkCampaignNotDraft(campaignId, res) {
```
Should we just default everything to zero for draft campaigns to remain consistent across the view Campaign views?

andioneto replied:
I spoke to V about this the other day and we decided that for campaigns still in draft this endpoint isn't accessible, but we can discuss that as a group since it's kind of a higher-level decision. @allieglotfelty

allieglotfelty commented on server/db/services/calls.js:
```diff
@@ -78,4 +78,13 @@ export default {
       .then(call => call)
       .catch(err => err);
   }
+  // TODO: remove if decided not to use
```
Should this be removed?

andioneto replied:
yesss

allieglotfelty commented on server/controllers/controller_helpers/metrics.js:
```diff
@@ -0,0 +1,115 @@
+    });
+}
+
+function createStatusMap(metricsObj) {
```
This is a design choice, but I'm wondering if it makes sense to abstract out the metricsObj rather than passing it into each of these helpers? Then add them to the metricsObj in countStatusAndOutcomeFrequency and handleQuestionResponses?

andioneto replied:
Yeah, I thought it was a bit more readable this way, instead of having assignments, the function invocations read as what they're doing. But I def see what you mean. I think it's just preference, do you see a performance benefit either way?

allieglotfelty replied:
I think it's written properly, but I suppose it's possible that unrelated part of the metricsObj could get updated by accident? Each function is only updating part of it, but we're passing the whole thing in, you know?

allieglotfelty replied:
I think in the interest of time we can leave it as is.

andioneto replied:
Ok, coo beans.

andioneto replied:
*cool

allieglotfelty commented on server/db/services/campaigns.js:
```diff
@@ -25,6 +25,13 @@ export default {
+  getCallsByCampaignId: (params) => {
+    const { id } = params;
+    return new Campaign({ id })
+      .fetch({ withRelated: 'calls' });
```
I _think_ you can also do .calls().fetch(), but this works too!

allieglotfelty commented on server/controllers/controller_helpers/metrics.js:
```diff
@@ -0,0 +1,115 @@
+    const responseOptions = splitResponses(responseOptionString);
+    createResponsesMap(responseOptions, question_id, metricsObj);
+
+    return responsesService.fetchResponesByQuestionCallId({ calls_array, question_id })
```
silly, but small typo in fetchRespon**ses**ByQuestionCallId

andioneto replied:
seriously, can't spell-type
lol

allieglotfelty commented on server/db/services/responses.js:
```diff
@@ -4,5 +4,16 @@ export default {
     const { call_id, question_id, response } = params;
     return new Response({ call_id, question_id, response }).save();
+  },
+  fetchResponesByQuestionCallId: (params) => {
```
see above

allieglotfelty commented on test/server/services/responses.js:
```diff
@@ -70,39 +94,92 @@ describe('Responses Service tests', () => {
-                    .then((contact) => {
-                      this.callSaveParams.contact_id = contact.attributes.id;
-                      usersService.saveNewUser(this.userSaveParams)
+                  this.callSaveParams = [];
```
this.callSaveParams is set as an empty object above - should that be deleted?

andioneto replied:
yes, I'll do that meow

allieglotfelty commented on test/server/services/responses.js:
```diff
@@ -70,39 +94,92 @@ describe('Responses Service tests', () => {
+              this.responseParams[2].id = thirdResponseModel.attributes.id;
+              const { question_id } = this.responseParams[0];
+              const callIdsArray = this.callSaveParams.map(call => call.id);
+              responsesService.fetchResponesByQuestionCallId({ question_id, calls_array: callIdsArray })
```
last typo fix

andioneto committed:
fixed requested changes:
spelling errors
unnecessary userparams and user creation in response service testing

andioneto committed:
fixed contact list service test bug

andioneto committed:
Merge branch 'master' of https://github.com/Twilio-org/phonebank into ao-metrics

allieglotfelty reviewed (dismissed):
Looks good!

andioneto committed:
Merge branch 'master' of https://github.com/Twilio-org/phonebank into ao-metrics

perfectlynormalbeast commented on server/controllers/controller_helpers/metrics.js:
```diff
@@ -0,0 +1,115 @@
+        }
+        return true;
+      }
+      return res.status(500).json({ message: 'problem with db query for campaign by id' });
```
This isn't necessarily a 500? It might mean a 404 and the campaign wasn't found? It's the error block for getCampaignById that would imply a 500-ish error.

andioneto replied:
ok cool beans

perfectlynormalbeast commented on server/controllers/controller_helpers/metrics.js:
```diff
@@ -0,0 +1,115 @@
+        if (status === 'draft') {
+          throw new Error('Campaign metrics not available for campaigns still in draft.');
+        }
+        return true;
```
We need to be consistent about what we're returning - a boolean or the value from a res.status
Also, even when someone tries to get stats for a draft campaign, we should probably return a 204 no content or something since we're implying that no stats are available.

andioneto replied:
got it, after reading about erros too, I decided to remove the try catch statement

andioneto committed:
Merge branch 'master' of https://github.com/Twilio-org/phonebank into ao-metrics

perfectlynormalbeast commented on server/controllers/metrics.js:
```diff
@@ -0,0 +1,34 @@
+  try {
+    checkCampaignNotDraft(campaignId, res);
+  } catch (e) {
+    return res.status(400).json({ message: e });
```
A 204 might be more appropriate unless we want client-side logic to ban metric queries for draft campaigns.

andioneto replied:
Yep, this is changed

perfectlynormalbeast commented on server/controllers/metrics.js:
```diff
@@ -0,0 +1,34 @@
+
+        return handleQuestionResponses(questionsArray, metrics, callIdMap, res);
+      })
+      .catch(err => res.status(400).json({ message: err }));
```
This is likely not a user error, so 400 is a bad choice.

andioneto replied:
500 it is

perfectlynormalbeast commented on server/controllers/controller_helpers/metrics.js:
```diff
@@ -0,0 +1,115 @@
+  .then(() => {
+    res.status(200).json({ message: 'metrics generated successfully', metricsObj });
+  })
+  .catch(err => res.status(404).json({ message: `Problem fetching responses by question id: ${err}` }));
```
A catchall error like this is more likely to be a 5xx

andioneto replied:
✔️

perfectlynormalbeast commented on server/db/services/responses.js:
```diff
@@ -4,5 +4,16 @@ export default {
     const { call_id, question_id, response } = params;
     return new Response({ call_id, question_id, response }).save();
+  },
+  fetchResponsesByQuestionCallId: (params) => {
```
nifty query!

perfectlynormalbeast commented on server/controllers/controller_helpers/metrics.js:
```diff
@@ -0,0 +1,115 @@
+      return countResponseFrequency(responseCollection, metricsObj, question_id);
+    })
+    .catch((err) => {
+      res.status(404).json({ message: `couldn't find responses by question id: ${err}` });
```
Bad form to mutate the status midpromise, especially when multiple promises may encounter this. Best to let the parent promise.all or something above that handle res mutation;

andioneto replied:
got it

perfectlynormalbeast commented on server/controllers/controller_helpers/metrics.js:
```diff
@@ -0,0 +1,115 @@
+    const { status, outcome } = call.attributes;
+    const statusStats = metricsObj.status_distribution;
+    const outcomeStats = metricsObj.outcome_distribution;
+    if (Object.prototype.hasOwnProperty.call(statusStats, status)) {
```
Not 100% sure, but just want to make sure this is the best way to do this check and that there isn't a less verbose way of accomplishing this? Applies to multiple lines in this file.

andioneto replied:
there prob is, i think in the interest of time I'm going to say I can attempt to get back to this later, or it will be covered in the controller clean up we're doing. Does that sound fair? Just bc this is blocking Duy

perfectlynormalbeast reviewed (changes requested):
Left a couple of questions and minor changes

andioneto committed:
revisions based on PR comments

nguyenaiden commented on test/server/services/campaigns.js:
```diff
@@ -84,18 +133,50 @@ describe('Campaign service tests', () => {
       this.responseParams = {
         response: 'I don\'t like paragraphs'
       };
+// >>>>>>> f2c9fb425671905fc957bfeb08cfb1df380622d5
```
Are these comments neccesary?

andioneto committed:
forgot conflict comments

nguyenaiden reviewed (approved):
Looks good to me! 👍

perfectlynormalbeast commented on server/controllers/metrics.js:
```diff
@@ -10,25 +10,33 @@ export default function getCampaignMetrics(req, res) {
+  return campaignsService.getCampaignById({ id: campaignId })
+    .then((response) => {
+      const { status } = response.attributes;
+      if (checkCampaignNotDraft(status).constructor === Error) {
```
Isn't a try-catch better at catching this error?

perfectlynormalbeast reviewed (commented):
Just one question.
